nixos/adguardhome: run --check-config before merging

CRTified opened the pull request:
To prevent merging configurations on different `schema_version`, we will now check the existing configuration, which has the side-effect of updating the `schema_version`.

Fixes #246461 as described in https://github.com/NixOS/nixpkgs/issues/246461#issuecomment-1667663556

Should work/merge independent of #247828, which changes the generation of the default config.

## Description of changes

Add an additional call to `adguardhome --check-config` to the `preStart`.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [X] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

CRTified commented:
Pinging @dmoonfire as creator of the original issue, as well.

bjornfor commented on nixos/modules/services/networking/adguardhome.nix:
```diff
@@ -150,6 +150,11 @@ in
+          # First run a schema_version update on the existing configuration
+          # This ensures that both the new config and the existing one have the same schema_version
+          # Note: --check-config has the side effect of modifying the file at rest!
+          ${pkgs.adguardhome}/bin/adguardhome -c "$STATE_DIRECTORY/AdGuardHome.yaml" --check-config
```
`${pkgs.adguardhome}/bin/adguardhome` -> `${lib.getExe adguardhome}`?

CRTified replied:
Also fixed the other occurrences. Tests pass for me on `x86_64-linux`.

rhoriguchi commented:
@ofborg eval

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1368

CRTified committed:
nixos/adguardhome: run --check-config before merging

To prevent merging configurations on different `schema_version`,
we will now check the existing configuration, which has the side-effect of
updating the `schema_version`.

Fixes #246461 as described in https://github.com/NixOS/nixpkgs/issues/246461#issuecomment-1667663556

CRTified commented:
@wegank Hope it's fine to ping you, as you have merged the related #247828. This PR closes a related, repeatedly reported issue with AdGuard Home.

wegank commented on nixos/modules/services/networking/adguardhome.nix:
```diff
@@ -167,8 +167,13 @@ in {
+          # First run a schema_version update on the existing configuration
+          # This ensures that both the new config and the existing one have the same schema_version
+          # Note: --check-config has the side effect of modifying the file at rest!
+          ${lib.getExe pkgs.adguardhome} -c "$STATE_DIRECTORY/AdGuardHome.yaml" --check-config
```
Should `pkgs.adguardhome` be `cfg.package` instead?

CRTified replied:
True. Give me a second.
